### PR TITLE
Added a CustomTerrainDebugOverlay

### PR DESCRIPTION
**Pull request by Mailaender**
A custom terrain visualizing debug overlay, see https://github.com/OpenRA/OpenRA/pull/10764#issuecomment-186585903

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+			if (name == CommandName && devMode != null)
+				devMode.ShowCustomTerrain ^= true;
+
+			if (arg.Contains("allcells"))
+				allCells ^= true;
+		}
+
+		public IEnumerable<IRenderable> Render(Actor self, WorldRenderer wr)
+		{
+			if (devMode != null && !devMode.ShowCustomTerrain)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs]: The `devMode != null` makes this unconditionally render the overlays in shellmaps and for observers, which is pretty rotten for polish.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs]: You should follow the example of TerrainGeometryOverlay instead.

**Comment by Mailaender** [comment]
Updated. This make the trait also properly optional.

File: OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	{
+		public readonly string Font = "TinyBold";
+
+		public object Create(ActorInitializer init) { return new CustomTerrainDebugOverlay(init.Self, this); }
+	}
+
+	class CustomTerrainDebugOverlay : IWorldLoaded, IChatCommand, IRender
+	{
+		const string CommandName = "debugcustomterrain";
+		const string CommandDesc = "Toggles the custom terrain debug overlay. Optional parameter: 'allcells'";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs]: This optional parameter doesn't appear to do anything.  Based on the name I would expect it to show the terrain type for all cells, not just the ones with custom types?

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs]: How odd.  When I tried this again it worked, I must have typed it wrong 5 or 6 times...

However, this just shows 255 colored in the color of the actual terrain type.  If you're not going to show the actual terrain type, then what is the use case of this?

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs]: Adding the changes requested tag to either remove this or explain the use case that justifies it.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs]: It is a left over from a previous implementation. We don't need it. Removed.

**Comment by pchote** [comment]
The terrain type numbers are generated by the engine and aren't exposed anywhere that a modder could find them.  This works a bit better (especially when zoomed) with the following change:
```diff
diff --git a/OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs b/OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs
index abc3d8e..71425fe 100644
--- a/OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs
+++ b/OpenRA.Mods.Common/Traits/Render/CustomTerrainDebugOverlay.cs
@@ -71,8 +71,8 @@ public IEnumerable<IRenderable> Render(Actor self, WorldRenderer wr)
                                if (terrainType == byte.MaxValue)
                                        continue;
 
-                               var color = wr.World.Map.GetTerrainInfo(cell).Color;
-                               yield return new TextRenderable(font, center, 0, color, terrainType.ToString());
+                               var info = wr.World.Map.GetTerrainInfo(cell);
+                               yield return new TextRenderable(font, center, 0, info.Color, info.Type);
                        }
                }
        }
```

screen shot 2016-06-05 at 00 01 00

**Comment by Mailaender** [comment]
Updated with friendly terrain names and explicit interfaces.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/DebugMenuLogic.cs [hunk 3]
```diff
@@ -134,6 +134,18 @@ public DebugMenuLogic(Widget widget, World world)
 				showActorTagsCheckbox.IsChecked = () => devTrait.ShowActorTags;
 				showActorTagsCheckbox.OnClick = () => devTrait.ShowActorTags ^= true;
 			}
+
+			var showCustomTerrainCheckbox = widget.GetOrNull<CheckboxWidget>("SHOW_CUSTOMTERRAIN_OVERLAY");
+			if (showCustomTerrainCheckbox != null)
+			{
+				var customTerrainDebugTrait = world.WorldActor.Trait<CustomTerrainDebugOverlay>();
+				showCustomTerrainCheckbox.Disabled = customTerrainDebugTrait == null;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/DebugMenuLogic.cs]: Checking for null indicates that you want `TraitOrDefault`.

Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/DebugMenuLogic.cs]: Yes, indeed. Thank you.

**Comment by GraionDilach** [comment]
:+1:

**Comment by reaperrr** [comment]
:+1:

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/038d2174c52c285788fa91b99d13bcb4e71614f5...95406a831475da885087de2a4f52b301437b53da)

**Pull request merged**